### PR TITLE
feat: Add support ed25519-signature-2020 for JSONLD

### PR DESCRIPTION
**Pull request by DaevMithran**
Signed-off-by: DaevMithran <daevmithran1999@gmail.com>

## What is being changed
This PR adds experimental support of Ed25519Signature2020 signature suite

## Issues faced

- The local key's publicKeyHex is different from resolved DID's publicKeyHex it fails at this [step](https://github.com/DaevMithran/veramo/blob/ed255192020-JSONLD/packages/credential-ld/src/action-handler.ts#L232)

> change suggested by @Eengineer1 : change encoding format from 'base16' [here](https://github.com/uport-project/veramo/blob/7b6d1950364c8b741dd958d29e506b95fa5b1cec/packages/utils/src/did-utils.ts#L296) to 'hex' [here](https://github.com/DaevMithran/veramo/blob/2b36fca2a9ca05b7f4a690b54807e173438b709d/packages/utils/src/did-utils.ts#L299) but this did not solve the issue

- Getting an invalid function error at this [step](https://github.com/DaevMithran/veramo/blob/ed255192020-JSONLD/packages/credential-ld/src/suites/Ed25519Signature2020.ts#L54) . Link to the [package](https://www.npmjs.com/package/@digitalcredentials/ed25519-signature-2020)

### CODE REVIEW

File: packages/credential-ld/src/suites/Ed25519Signature2020.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import { encodeJoseBlob } from '@veramo/utils'
+import { RequiredAgentMethods, VeramoLdSignature } from '../ld-suites'
+import { CredentialPayload, DIDDocument, IAgentContext, IKey, TKeyType } from '@veramo/core'
+import * as u8a from 'uint8arrays'
+// @ts-ignore
+import { Ed25519VerificationKey2020, Ed25519Signature2020 } from '@digitalcredentials/ed25519-signature-2020'
```
Comment (Eengineer1) [inline packages/credential-ld/src/suites/Ed25519Signature2020.ts]: @DaevMithran Please resolve this and remove `@ts-ignore`.

File: packages/credential-ld/src/suites/Ed25519Signature2020.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      },
+    }
+
+    const verificationKey = await Ed25519VerificationKey2020.from({
+      id,
+      controller,
+      publicKeyMultibase: u8a.fromString(key.publicKeyHex, 'base16'),
+      signer: ()=> signer,
+      type: this.getSupportedVerificationType(),
+    })
```
Comment (Eengineer1) [inline packages/credential-ld/src/suites/Ed25519Signature2020.ts]: @DaevMithran Can you fix instantiations as well?
This is incorrect.
You can utilise the class structure to replace this block, as this is brought over from `Ed25519Signature2018`.

**Review by Eengineer1 — CHANGES_REQUESTED** [review]
@DaevMithran Needs polishing & clean-ups, but it's heading to a good direction overall.
Can you also provide a short summary on how you're testing your approach / your local setup for testing and potentially changes you're making on tools, e.g. context loaders, resolver?

Let me know when you've made the changes requested as well.

**Comment by DaevMithran** [comment]
Hey @Eengineer1 

I have pushed changes. I was able to create a JSONLD credential after these changes

1. Change publicKeyMultibase encoding [here](https://github.com/DaevMithran/veramo/blob/ed255192020-JSONLD/packages/credential-ld/src/suites/Ed25519Signature2020.ts#L50): The publicKey should be encoded using `MULTICODEC_ED25519_PUB_HEADER = new Uint8Array([0xed, 0x01])` This should be done in the cheqd/sdk created a pull request for that [here](https://github.com/cheqd/sdk/pull/43).
2. Modified the signer's anonymous sign funciton [here](https://github.com/DaevMithran/veramo/blob/ed255192020-JSONLD/packages/credential-ld/src/suites/Ed25519Signature2020.ts#L35)
3. Declared the digitalcredenitals package modules [here](https://github.com/DaevMithran/veramo/blob/ed255192020-JSONLD/packages/credential-ld/src/module-types/jsonld/index.d.ts)
4. Added an assertion method to the didDoc [here](https://github.com/DaevMithran/veramo/blob/ed255192020-JSONLD/packages/utils/src/did-utils.ts#L161). This also needs to be done in cheqd/sdk pushed this change in the same pull request mentioned above


Further Improvements

- the signatureMap function allows only one suit for a KeyType [link](https://github.com/DaevMithran/veramo/blob/ed255192020-JSONLD/packages/credential-ld/src/ld-suite-loader.ts#L10) . Both Ed25519Signature2020 & Ed25519Signature2018 have the same KeyType: Ed25519. I can change KeyType as Ed255192018. It will work

**Review by Eengineer1 — CHANGES_REQUESTED** [review]
@DaevMithran It seems the `cheqd:feat/jsonLDSupport` branch was not in sync with the respective upstream branch, resulting in a total mess of commits included with your personal changes.
I've synchronized it, so you're good to go.

Steps to resolve this:
- Switch to `cheqd` Veramo fork first (next time please do start from here, rather than your personal `DaevMithran` fork <-- this one you only control yourself, so redundant complexity).
- Carry over your changes to `feat/jsonLDSupport`.
- Target that PR to `cheqd:next`.
- Close *this* one.
- On internal approval, we can open a final PR for upstream review by targeting `veramo:next`.

Thanks!

**Comment by DaevMithran** [comment]
@Eengineer1 sorry didn't notice this. Can I fork the veramo's next branch and push it in cheqd/jsonLdsupport. 

By that way it will be easier to create a pull request to veramo's next branch. I think the other commits in cheqd's next branch are irrelevant

**Comment by Eengineer1** [comment]
@DaevMithran 
> I think the other commits in cheqd's next branch are irrelevant

That's correct, but that's the exact reason for the suggested taxonomy.

Could you try following the described steps above?
It will be easier to do a thorough review as a result.

**Pull request closed without merging**